core/client/src/main/java/alluxio/client/file/FileOutStream.java

```diff
@@ -117,53 +123,46 @@ public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContex
-          mUfsFileId =
-              mFileSystemWorkerClient.createUfsFile(new AlluxioURI(mUfsPath),
-                  CreateUfsFileOptions.defaults().setPermission(perm));
-          mUnderStorageOutputStream = mUnderOutStreamFactory
-              .create(mFileSystemWorkerClient.getWorkerDataServerAddress(), mUfsFileId);
-        } catch (AlluxioException e) {
-          mFileSystemWorkerClient.close();
-          throw new IOException(e);
-        } catch (IOException e) {
-          mFileSystemWorkerClient.close();
-          throw e;
-        }
-      } else {
-        String tmpPath = PathUtils.temporaryFileName(mNonce, mUfsPath);
-        UnderFileSystem ufs = UnderFileSystem.get(tmpPath);
-        // TODO(jiri): Implement collection of temporary files left behind by dead clients.
-        CreateOptions createOptions = new CreateOptions().setPermission(options.getPermission());
-        mUnderStorageOutputStream = ufs.create(tmpPath, createOptions);
-
-        // Set delegation related vars to null as we are not using worker delegation for ufs ops
-        mFileSystemWorkerClient = null;
-        mUfsFileId = null;
-      }
-    } else {
+    mClosed = false;
+    mCanceled = false;
+    mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
+    mBytesWritten = 0;
+    if (!mUnderStorageType.isSyncPersist()) {
       mUfsPath = null;
       mUnderStorageOutputStream = null;
       mFileSystemWorkerClient = null;
       mUfsFileId = null;
+    } else {
+      try {
+        // Get the ufs path from the master.
+        try (CloseableResource<FileSystemMasterClient> client = mContext
+            .acquireMasterClientResource()) {
+          mUfsPath = client.get().getStatus(mUri).getUfsPath();
+        }
+        if (mUfsDelegation) {
+          mFileSystemWorkerClient = mCloser.register(mContext.createWorkerClient());
+          Permission perm = options.getPermission();
+          mUfsFileId = mFileSystemWorkerClient.createUfsFile(new AlluxioURI(mUfsPath),
+              CreateUfsFileOptions.defaults().setPermission(perm));
+          mUnderStorageOutputStream = mCloser.register(mUnderOutStreamFactory
+              .create(mFileSystemWorkerClient.getWorkerDataServerAddress(), mUfsFileId));
+        } else {
+          String tmpPath = PathUtils.temporaryFileName(mNonce, mUfsPath);
+          UnderFileSystem ufs = UnderFileSystem.get(tmpPath);
+          // TODO(jiri): Implement collection of temporary files left behind by dead clients.
+          CreateOptions createOptions = new CreateOptions().setPermission(options.getPermission());
+          mUnderStorageOutputStream = mCloser.register(ufs.create(tmpPath, createOptions));
+
+          // Set delegation related vars to null as we are not using worker delegation for ufs ops
+          mFileSystemWorkerClient = null;
+          mUfsFileId = null;
+        }
+      } catch (AlluxioException | IOException e) {
+        mCloser.close();
```

Comment:
good question. because we don't want `mCloser` to close unless there is exception thrown. The convention I made is in constructor, I only put `mCloser.close()` in exception branch.